### DISCUSSION TITLE
Composite Time Window Partitions Definitions

### DISCUSSION THREAD

**Discussion by cbini** [dagster-io/dagster#11809] [Ideas]
I've been grappling with this issue, and I think a new feature might be the solution. 

For one of our systems, we'd like to materialize assets hourly using a partition. However, some of this data can go as far back as 2002, which results in an insane number of partitions.

It would be extremely helpful if we could define a sets of partition types that use start and end dates as boundaries. That way you could define large time windows for historical data and smaller time windows for more-recent and future data.

I imagine it would looks something like this:

```
historical_partitions_def = TimePartitionsDefinition(
    start_date="2002-07-01T00:00:00.000000-04:00",
    end_date="2022-06-30T23:59:59.999999-05:00",
    timezone=str(LOCAL_TIME_ZONE),
    cron_schedule="0 0 1 7 *",
    fmt="%Y-%m-%dT%H:%M:%S.%f%z",
)

daily_partitions_def = DailyPartitionsDefinition(
    start_date="2022-07-01T00:00:00.000000-05:00",
    end_date="2023-01-18T23:59:59.999999-04:00",
    timezone=str(LOCAL_TIME_ZONE),
    fmt="%Y-%m-%dT%H:%M:%S.%f%z",
)

hourly_partitions_def = HourlyPartitionsDefinition(
    start_date="2023-01-19T00:00:00.000000-04:00",
    timezone=str(LOCAL_TIME_ZONE),
    fmt="%Y-%m-%dT%H:%M:%S.%f%z",
)

@asset(
    name="my_asset_table",
    partitions_def=[historical_partitions_def, daily_partitions_def, hourly_partitions_def]
)
def my_asset_table(context):
    ...
```

**Comment by sryza** [comment]
Hey @cbini . I can't promise that we'll be able to get to it soon, but I think this is a reasonable approach.

I'm curious - do you ever backfill the data all way way back from the beginning?

**Reply by cbini** [reply]
I tried to using partitions in a straightforward way, but backfilling even a single table created a ton of runs that overloaded our cluster.

To make it work, I cheated by setting the start date to yesterday and if the partition key matched the start date, the query adjusted to select all records less than the partition date.

That way, I have all our historical data in the bank but moving forward, it will update incrementally each hour.

**Comment by sryza** [comment]
Also, if you don't mind, I changed the name of this discussion to be a little more specific. There's an ancient thing in Dagster called "partition sets", which is basically another name for a PartitionedConfig, so the original title was throwing me off.

**Comment by sryza** [comment]
Here's a relevant request: 
> I have an asset for which I can fetch a daily file all the way up to 2021. Before that, the same data is provided in a zip file quarterly. So the whole 2020 daily data is provided in 4 zip files. I can create 2 separate assets, the first using DailyPartitionsDefinition and the second using StaticPartitionsDefinition but how can I "hide" this implementation detail in downstream assets? Is it possible to "merge" these 2 assets into a single one that spans the whole time range using DailyPartitionsDefinition?

https://dagster.slack.com/archives/C01U954MEER/p1677535394731719
